Add performance test

iomartin opened the pull request:
This script sets up a RAID5 MD array and optionally runs an fio job
against it, capturing the BW, IOPS and CPU utilization.

CLI options allow setting the MD chunk size, group_thread_cnt and
stripe_cache_size, as well as the fio block size, IO pattern, numjobs,
iodepth and runtime.

It is also possible to pass a list of devices to be used for the MD
array. If no devices are passed, the script uses all unmounted
non-Eideticom NVMe devices.

iomartin committed:
perf_test: add performance test script

This script creates an MD array and measures its performance, namely
sequential throughput, random IOPS and CPU utilization.

It also supports testing different configurations using different CLI
options: array chunk size, number of threads and cache size, as well as
fio blocksize.

It prints a summary at the end with the top result for each
metric/blocksize.

lsgunth commented on md_test.py:
```diff
@@ -0,0 +1,168 @@
+    devs = json.loads(devs)
+    devs = filter(lambda d: "Eideticom" not in d["ModelNumber"], devs["Devices"])
+    devs = [d["DevicePath"] for d in devs]
+    debug(f"Non-Eideticom devices found: {' '.join(devs)}")
```
This is a public repo.... Not sure I like this Eideticom specific noise creeping out... Is there another way we can handle this?

iomartin replied:
Maybe add a `--skip-devices` option so it uses all devices it finds except those in the list?

lsgunth commented on md_test.py:
```diff
@@ -0,0 +1,168 @@
+
+def debug(*args):
+    if verbose:
+        print(">", *args)
```
It might be better to use python logging instead of reinventing debug prints.... And the ugly global variable....

lsgunth commented on md_test.py:
```diff
@@ -0,0 +1,168 @@
+           "--run", "--assume-clean", f"--raid-devices={len(devs)}"] + devs
+    run(cmd=cmd, stderr=sp.DEVNULL)
+
+    with open(f"/sys/block/{MD}/md/group_thread_cnt", "w") as f:
```
I rather prefer pathlib for interfacing with sysfs.... path.read_text() and path.write_text() is a bunch nicer and you can do things like:

```python
sysfs_block = pathlib.Path("/sys/block")

(sysfs_block / MD / "md" / "group_thread_cnt").write_text(group_thread_cnt)
```

lsgunth commented on md_test.py:
```diff
@@ -0,0 +1,168 @@
+             stripe_cache_size=args.stripe_cache_size)
+
+    if not args.setup_only:
+        run_fio(bs=args.block_size, rw=args.readwrite, numjobs=args.numjobs,
```
Seems like we want this script to summarize and output a bunch of the metrics we care about:  read and write (separately) for sequential IO bandwidth and random IOPS. And find some optimum parameters for a fast machine.

A script that just takes in a ton of parameters for a single fio run isn't quite what we need.

dlsloan replied:
Might be overkill for this but we do have an fio python script for getting performance numbers already in: [linux-noload:fio.py](https://github.com/Eideticom/linux-noload/blob/master/tools/testing/selftests/noloadfs/performance/fio.py)
Don't need all the tempfile stuff when running on a block device, but it runs fio scripts passing fio arguments down in a pythonic way "fio(blocksize=...)" and returns the json formatted fio output.

lsgunth reviewed (commented):
Also can we call it perf_test or something related to "performance"?

dlsloan commented on md_test.py:
```diff
@@ -0,0 +1,168 @@
+import sys
+import time
+
+MD = "md0"
```
if we're going through all the existing md devs anyway why not just pick an unused md dev then?

lsgunth replied:
We're going through all the nvme devices not the md devs.... Probably not worth the effort. All the other test infrastructure assumes nothing else is using md on start.

iomartin commented:
Force-pushed rebasing on #2.

The new version can take as input a list of chunk sizes, cache sizes, thread cnt and fio blocksizes and tries all different configurations.

It can measure sequential bw, random iops and cpu utilization, for both read and write and there are knobs to measure just some of these if required.

dlsloan commented on perf_test:
```diff
@@ -0,0 +1,165 @@
+
+    # Remove mounted devices
+    with open("/proc/mounts") as f:
+        mounts = " ".join(f.readlines())
```
While this probably works most of the time, things with path args or stacked filesystems could get messy, 
something like:
```
for line in f:
  dev, mount, fs, opts, dump, pass = line.split()
  #dev check ...
```
Actually checks the correct argument

lsgunth replied:
Yeah, I kind of think this is too clever by a half as well... Maybe a config file so the user can explicitly specify which devices to use or some such. Then we just need to maintain a config file on specific machines.

iomartin commented:
I removed the device search option.

I explored using a config file with configparser but it felt like more trouble than it was worth if we wanted to make it play nicely with argparse. It's simpler to just create wrappers with the disks and whatever options are required.

iomartin commented:
I reworked this so that it accepts all the args from md.py rather than keeping its own list.

I also rebased on top of #3, so only the last 5 commits are part of this PR.

dlsloan commented on md.py:
```diff
@@ -156,7 +156,7 @@ def setup(self, level=5, devs=None, ndisks=None, disk_type='ram',
             (self._sysfs / "stripe_cache_size").write_text(str(cache_size))
 
-    def parse_args(self, args=None):
+    def parse_args(self, args=None, extra_args=[]):
```
I feel like we should be pulling the arg-parser out of md.py and instead expose a way to register it's args with whatever script is calling it rather than making every script use md.py's parse_args, this is all a bit messy to me and we're adding a new interface to remember, rather than exposing argparse features directly.

lsgunth replied:
Yeah, I don't like the extra_args method much either, but I can see why Martin did it this way. 

Try something like this:

```python3

class MDArgumentParser(EnvironmentArgumentParser):
    def __init__(self):
         super().__init__()
         grp = self.add_argument_group("md creation arguments")
         grp.add_argument(...)
         ...

def setup_from_parsed_args(self, parsed_args):
    ...

def setup_from_args(self, args=None):
     parser = MDArgumentParser()
     args = parser.parse_args(args)
     self.setup_from_parsed_args(self, args)
```

iomartin replied:
Good ideas. I've pushed a new version that implements MDArgumentParser

iomartin commented:
I've also changed the PR target branch to [[dlsloan/nvme]](https://github.com/Eideticom/raid5-tests/tree/dlsloan/nvme) as it depends on a few things from it and then it's easier to see which commits are relevant to this PR.

dlsloan reviewed (approved):
My gut says we don't really need a new class for the MDParser stuff, but that might be getting a bit pedantic, and the rest looks good to me :)

lsgunth commented on md.py:
```diff
@@ -6,6 +6,27 @@
+            return False
+    return False
+
+def env_action(action):
```
I'd much prefer these to remain part of EnvironmentArgumentParser.... It's silly that they are outside of that scope.... This is all kind of unfortunate, but I may take a closer look at this tomorrow. I didn't realize the code I
copied made use of single underscore stuff and wouldn't mind finding a better way.

lsgunth commented on md.py:
```diff
@@ -18,27 +39,57 @@ def __init__(self, *, formatter_class=_CustomHelpFormatter,
+    def add_argument_group(self, *args, **kwargs):
+        grp = super().add_argument_group(*args, **kwargs)
+        grp.__class__ = EnvironmentArgumentGroup
+        return grp
```
This should be part of EnvironmentArgumentParser not MDArgumentParser.

lsgunth commented on md.py:
```diff
@@ -49,13 +49,15 @@ def _add_action(self, action):
 
 class MDArgumentParser(EnvironmentArgumentParser):
-    def __init__(self):
+    def __init__(self, md_nargs=1):
```
I don't like this much at all. I think it'd be better to just replace these arguments in the perf script like this:

```python3
parser = md.MDArgumentParser(conflict_handler='resolve')
parser.add_argument("-c", "--chunk-size", default=[64 << 10],
                    type=parser._suffix_parse, nargs="+",
                    help="md chunk size (specify multiple arguments to sweep)")

args = parser.parse_args()
```

(Note the conflict_handler in the parser instantiation)

But maybe add a commit to remove the single underscore on suffix_parse that I added.

iomartin replied:
Yeah, I didn't like it much either, the conflict_handler is way nicer

iomartin committed:
md.py: convert list env vals to correct type

Otherwise they'd always be a list of strings.

iomartin committed:
md.py: save common argument group

For scripts that want to add/overwrite arguments in the common group.

iomartin commented:
Changed the target branch back to main

lsgunth reviewed (commented):
This looks good to me once David's PR is in. 

Also, interested in seeing results first though.

dlsloan commented:
@iomartin you should get this merged before I break it again with my changes ;)

iomartin commented:
I changed how this works substantially and rewrote a bunch of stuff.

First, I dropped the `--blocksize` argument and instead allowed `--bandwidth` and `--iops` to take arguments so that you can specify different blocksizes for each measurement.

Second, I added a summary table to the end that shows the best performance for a given blocksize and the corresponding parameters.

Finally, I split the big table into up to 4 narrower tables (write BW, read BW, write IOPS, read IOPS). @lsgunth would you still prefer if these were written to a separate file instead of printed to stdout?

For example (disregard the actual measurements since I ran the fio for just 1s):

```
$ sudo ./perf_test --devs /dev/nvme{1..7}n1 --disks 7 --runtime 1 --quiet --run --disks 7 \
                   --cpu --chunk-size 64K 128K --cache-size 4096 8192 \
                   --iops 4K 8K --bw 1M 2M --read --write
Seq Write BW
blocksize      chunk_size    thread_cnt    cache_size  Seq Write BW    CPU User    CPU System
-----------  ------------  ------------  ------------  --------------  ----------  ------------
1M                  65536             4          4096  5.35 GiB/s      2.8%        15.5%
1M                  65536             4          8192  5.27 GiB/s      3.0%        16.5%
1M                 131072             4          4096  4.15 GiB/s      2.7%        14.9%
1M                 131072             4          8192  3.51 GiB/s      2.3%        17.6%
2M                  65536             4          4096  4.51 GiB/s      2.3%        20.8%
2M                  65536             4          8192  3.94 GiB/s      2.3%        39.3%
2M                 131072             4          4096  4.35 GiB/s      2.5%        16.1%
2M                 131072             4          8192  4.01 GiB/s      2.4%        21.0%

Seq Read BW
blocksize      chunk_size    thread_cnt    cache_size  Seq Read BW    CPU User    CPU System
-----------  ------------  ------------  ------------  -------------  ----------  ------------
1M                  65536             4          4096  13.2 GiB/s     1.7%        3.9%
1M                  65536             4          8192  13.2 GiB/s     1.6%        3.9%
1M                 131072             4          4096  13.2 GiB/s     1.6%        3.4%
1M                 131072             4          8192  10.2 GiB/s     1.7%        2.6%
2M                  65536             4          4096  12.9 GiB/s     1.6%        4.0%
2M                  65536             4          8192  12.8 GiB/s     1.4%        4.0%
2M                 131072             4          4096  10.4 GiB/s     1.5%        3.1%
2M                 131072             4          8192  13.2 GiB/s     1.4%        3.3%

Rnd Write IOPS
blocksize      chunk_size    thread_cnt    cache_size    Rnd Write IOPS  CPU User    CPU System
-----------  ------------  ------------  ------------  ----------------  ----------  ------------
4K                  65536             4          4096             93106  2.1%        8.3%
4K                  65536             4          8192             92262  2.1%        8.3%
4K                 131072             4          4096             99017  2.3%        8.3%
4K                 131072             4          8192            101247  2.4%        10.0%
8K                  65536             4          4096             69741  1.9%        10.4%
8K                  65536             4          8192             64401  2.0%        10.0%
8K                 131072             4          4096             68755  1.9%        8.8%
8K                 131072             4          8192             68674  2.0%        10.0%

Rnd Read IOPS
blocksize      chunk_size    thread_cnt    cache_size    Rnd Read IOPS  CPU User    CPU System
-----------  ------------  ------------  ------------  ---------------  ----------  ------------
4K                  65536             4          4096           327599  3.1%        4.5%
4K                  65536             4          8192           325678  2.8%        5.0%
4K                 131072             4          4096           324706  3.0%        4.7%
4K                 131072             4          8192           330967  3.1%        4.6%
8K                  65536             4          4096           274832  2.8%        4.6%
8K                  65536             4          8192           279728  3.0%        4.7%
8K                 131072             4          4096           272249  3.0%        4.4%
8K                 131072             4          8192           275776  3.1%        4.5%

Best configurations:
Metric          blocksize    Value         chunk_size    thread_cnt    cache_size  CPU User    CPU System
--------------  -----------  ----------  ------------  ------------  ------------  ----------  ------------
Seq Write BW    1M           5.35 GiB/s         65536             4          4096  2.8%        15.5%
Seq Write BW    2M           4.51 GiB/s         65536             4          4096  2.3%        20.8%
Seq Read BW     1M           13.2 GiB/s         65536             4          4096  1.7%        3.9%
Seq Read BW     2M           13.2 GiB/s        131072             4          8192  1.4%        3.3%
Rnd Write IOPS  4K           101247            131072             4          8192  2.4%        10.0%
Rnd Write IOPS  8K           69741              65536             4          4096  1.9%        10.4%
Rnd Read IOPS   4K           330967            131072             4          8192  3.1%        4.6%
Rnd Read IOPS   8K           279728             65536             4          8192  3.0%        4.7%
```

lsgunth reviewed (approved):
Good enough. Looks good to me.

dlsloan reviewed (approved):
Looks good to me
